### DISCUSSION TITLE
issue: nested objects controlled by usefieldarray values are set to `true` onblur for empty string values

### DISCUSSION THREAD

**Discussion by jogly** [react-hook-form/react-hook-form#6835] [Q&A] [answered]
### Version Number

7.17.2

### Codesandbox/Expo snack

https://codesandbox.io/s/proud-cache-cp5dp?file=/src/Form.tsx

### Steps to reproduce

Its important to have `<DevTool />` in the reproduction sandbox.  Without this component, this bug disappears and the nested objects fields are correctly left as `""`.  However, the root cause is not the devtool because I encounter this issue in my app's code without DevTool included.  Unfortunately, I cannot determine what specifically DevTool does that causes this, I can only guess that it has to do with `watch`ing the field array?

1. Add item to field array by clicking `Add item` button.  _Note, it seems important that the nested object is not set (either `undefined` or `null`) when adding the item to the array.  In my production code I tried defining the nested object on `append` but that didnt resolve this bug.  However, I cannot reproduce the bug when the nested object is not nullish in the minimal reproduction, so that might be my fault or there is another way to trigger this issue besides having the field set non-nullish on `append`._
2. Define the nested object by clicking `Add nested item` button.
3. Delete the text in the second `input` with text `level 2a`.  _At this point, the value is correctly set to `""`, as seen in the DevTool and console logs._
![image](https://user-images.githubusercontent.com/1476969/137197103-5d3dae50-1611-4011-83aa-85ed97616bfd.png)
![image](https://user-images.githubusercontent.com/1476969/137197129-094f8ce4-6c21-4822-b2ae-2d42800a5c1b.png)

4. Unfocus the text input, and notice in the DevTool the value is now `true`.
![image](https://user-images.githubusercontent.com/1476969/137197182-1b19e80d-3265-47e0-8baf-4831d4566617.png)

5. Change the text in the first `input` field to trigger a re-render and notice the value for `items[0].nested.a` in the console log is `true`
![image](https://user-images.githubusercontent.com/1476969/137197219-6fba973e-0496-4624-a7db-bf027b9db3ae.png)

### Expected behaviour

When a nested objects field input value is `""`,  I expect the value to stay `""` when unfocusing the input.

I'd like to acknowledge the rule for `useFieldArray`:

> When you append, prepend, insert and update the field array, the obj can't be empty object rather need to supply all your input's defaultValues.

I may have interpreted incorrectly, but it seemed to me that this rule doesn't prohibit having object fields being `undefined` or `null` (e.g. `append({"primitive": "level 1", "nested": null})`) if I intend to omit the relevant nested `input`s.  It also seemed like a bug due to the behavior being expected without `<DevTool />` and unexpected when included.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
> When you append, prepend, insert and update the field array, the obj can't be empty object rather need to supply all your input's defaultValues.

it's important to provide what's to append/prepend for useFieldArray, as the hook is uncontrolled by nature, so you will need to supply the inputs values for the action.

```tsx
<button onClick={() => append({ primitive: "level 1", nested: null })}>
<button onClick={() => append({ primitive: "level 1", nested: undefined })}>
 // doesn't help useFieldArray to figure out what to popuatle for your inputs dyanmtically
```

it's also important to maintain the shape of your inputs, in your example is different

`null vs { a: string, b: string}`

**Comment by jogly** [comment]
> `// doesn't help useFieldArray to figure out what to popuatle for your inputs dyanmtically`

Right, but I don't expect any`<input .../>` to exist if `nested` is undefined, as there's a ternary in the example:
```tsx
data.items?.[index]?.nested ? (
              <input
                defaultValue=""
                key={`${index}.nested.a`}
                {...register(`items.${index}.nested.a` as const)}
              />
            ) : //....
```

Once I call `update(index, { ...data?.items?.[index], nested: { a: "level 2a", b: "level 2b" }  })`, I am providing the shape of the nested object, which causes a re-render, *then* I register the field.
You say

> it's also important to maintain the shape of your inputs, in your example is different
> `null vs { a: string, b: string}`

which I think may be more pertinent but I dont know why this is important.  I must be missing something fundamental, sorry if this is an obtuse question: why must the shape be maintained if I'm not registering any inputs before they exist?

edit: thank you for your time still, I can't believe your response time on a package this popular.  I know this is a crazy edge case, I thought it might be worth sharing bc it cost me a lot of time figuring out why my data was `true` sometimes 😄

**Comment by bluebill1049** [comment, marked as answer]
The reason is react hook form `useForm` is uncontrolled by default (for the entire form). 

## Uncontrolled form

we don't need anything because it falls back to an empty string (native form)

```tsx
<form>
  <input {...register('test')} /> 
</form>
```

## Controlled form: 

```tsx
useForm({
  defaultValues: {test: 'test'} // this is required to have 
})
const formValues = watch();

<form>
  <input onChange={(e) => setValue('test', e.target.value)}  value={formValues.test} />
</form>
```

and also individually controlled components (Controller). 

```tsx
useForm({
  defaultValues: {test: 'test'} // this is required to have 
})
const { field } = useController();

<input {...field} />
```

----------------------------------------------------------------

This applies to the `useFieldArray`

### What if i `append()` with nothing?

because `useFieldArray` is uncontrolled, it would have no clue what's the `defaultValue` would be for those dynamically append inputs, for example, the default for an input/controlled input could it be `undefined`, `null`, `''` or even `0` in some cases. So prediction is not going to work out, and that's why `useFieldArray` needs to know what's the defaultValue for each input when they mounted. 

````tsx
// ok now we know what's the defaultValue to be appended when each inputs mounted, no guessing game.
append({
  firstName: '',
  lastName: '',
})
````

hope the above make more sense.

-----------------

the code below:

```tsx
update(index, {
  ...data?.items?.[index],
  nested: { a: "level 2a", b: "level 2b" }
})
```

in fact, in your example

```diff
- <button onClick={() => append({ primitive: "level 1", nested: null })}>
+ <button onClick={() => append({ primitive: "level 1", nested: { a: '', b: ''} })}>
  add item
</button>
```

will result in the correct behavior

**Comment by jogly** [comment]
Thanks for the additional detail!  Unfortunately it doesn't really get to the point that I'm asking about.  I understand when any values are undefined / null, theres absolutely no way for this library to predict the default values.  No problem at all, because no inputs are mounted in the DOM that reference a path that is undefined.

The scenario I'm reporting in this issue is when the default values are specified later, but still before an input is mounted: 

```tsx
update(index, {
  ...data?.items?.[index],
  nested: { a: "level 2a", b: "level 2b" }
})
```

In this case, the default values are defined before the input elements are mounted:
```tsx
data.items?.[index]?.nested ? (
              <input
                defaultValue=""
                key={`${index}.nested.a`}
                {...register(`items.${index}.nested.a` as const)}
              />
            ) : //....
```

**Comment by bluebill1049** [comment]
how can I reproduce the update issue? 

the code below:

```tsx
update(index, {
  ...data?.items?.[index],
  nested: { a: "level 2a", b: "level 2b" }
})
```

in fact, in your example

```diff
- <button onClick={() => append({ primitive: "level 1", nested: null })}>
+ <button onClick={() => append({ primitive: "level 1", nested: { a: '', b: ''} })}>
  add item
</button>
```

will result the correct behaviour
